nixosTest matrix (prev also modularity, option docs)

roberth opened the pull request:
###### Description of changes

Use the module system to replace the NixOS test framework's custom glue.

 - NixOS tests can now use `imports` and other module system features.
 - Type checking for the test arguments.
 - Tests can now define overrides for interactive use.
 - Network-wide test setup can now be reusable; example:
     - A setup with all VMs configured to use a DNS server
     - Split long, slow tests into multiple tests that import a
       common module that has most of the setup.
 - We now have a module for test matrices / parameterized tests. We can stop reinventing the wheel, which we've done in many ways.
   - Generic support for version matrix + package passthru pattern (TBD)
 - (TBD) "generated" options reference docs
 - Implementation cleanup: aspects that had to be wired through all the glue are now in their own files.
    - `testing/interactive.nix`: just a couple of lines that can be ignored by the other modules, as opposed to wiring it through what is now `testing/run.nix`.
    - `testing/network.nix`, etc
 - For supporting containerized tests, the flexibility of the module system will be useful.

<!-- 
Furthermore, this adds `matrix` options for producing build matrices ergonomically.

... and when I say matrix, I do mean Matrix. I might tone down the whimsicality a bit, let me know what you think. At this time, I think the added imagery is amusing and refreshing, but I'm 100% biased.
-->

All in all, this quite significantly collapses the number of layers and custom expressions that test authors have to deal with. 



Also, more info in the expanded commit messages.

TODO:

 - [ ] clean up commit history
 - [ ] self-review the docs; make sure suggestions were applied correctly while rebasing
 - [x] AVOIDED: https://github.com/NixOS/nixpkgs/pull/177576
 - [x] https://github.com/NixOS/nixpkgs/pull/177157
 - [x] #163617 
    - `defaults` option
 - [x] do something about the now-legacy code that's duplicated
    - ~either move to "legacy" directory and leave in place until all migrated?~
    - ~or big bang and monitor the fallout~
    - use `nixos-test-staging` jobset in hydra
      - before rebase
         - latest _complete_ comparison https://hydra.nixos.org/eval/1768916?compare=1768212&full=0
      - baseline eval 2 https://hydra.nixos.org/eval/1769103
         - first comparison of 2nd rebase https://hydra.nixos.org/eval/1769104?compare=1769103&full=0
 - [x] maybe not put `name` in `specialArgs`; feasible with some `network.nix` refactoring to make the node `config` skeleton static.
 - [x] convert a couple more tests
        done:
          - `_3proxy`
          - `acme`
          - `adguardhome`
          - `aesmd`
          - `agate`
          - `avahi`
          - `cassandra`
          - `oci-containers`
          - `hostname`
          - `networking`
 - [x] docs in the manual
 - [x] do a real example of multi-dimensional matrix; have tested this manually but example was not commit-worthy
     - [x] test with variant that has extra configuration: `avahi`
     - [x] test with parameter: `oci-containers`
     - [x] demonstrate 2+ orthogonal choices (`networking`: `implementation` and `testcase` axes)
     - [ ] (unnecessary) demonstrate how to prune irrelevant choices (`matrix.<choice>.enable = false` + `matrix.<choice>.after`)
        - [x] with enable (`cassandra.passthru.tests.nixos`)
        - [ ] (unnecessary) with nesting (if that works) 
 - [x] add an example that modifies the nixos config directly from `matrix` (as `value.<value>` is a "parallel universe" of the entire test config) -> `avahi`
 - [x] `matrix`: rename `value` to `module`?
 - [x] #179351 
 - [x] generated docs

Closes #144110

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth committed:
nixos/testing/runTest: Add matrix options

roberth committed:
nixosTests.hostname: Convert to module and plug into the matrix

roberth committed:
nixosTests.hostname: Move initialization to all-tests.nix

This should be possible to do for all tests, so we can
remove boilerplate glue and replace it by modules and
matrix settings.

roberth committed:
nixos/testing/runTest: Add params option

roberth committed:
nixosTests.hostname: Simplify

mweinelt commented on nixos/lib/testing/matrix.nix:
```diff
@@ -0,0 +1,132 @@
+
+        you test in wonderland, and I show you how deep the rabbit hole goes...
+
+        Remember, all I'm offering is the multiverse... Follow me...
```
Just :heart:

fricklerhandwerk replied:
This is all fun, but should rather go to your personal blog post. Literally everyone outside the western cultural circle in their mid-30s will be completely thrown off. And possibly pissed to have to filter out what matters if they just want to get their job done.

(Please leave it in the Git history though, because it *is* awesome!)

roberth replied:
I agree. I already felt it was on the edge of acceptable, but you're absolutely right about the biases. Thank you for reminding me!
In a way it was also "hiding" the fact that that part of the documentation was rather incomplete. 

I haven't rebased it away, so you'll still be able to find it. (`Remove movie references`; hash might change)

roberth committed:
nixosTests.avahi: Use module system based runner

roberth committed:
nixos/testing/matrix: Add 0-cost extend to result

roberth committed:
nixosTests.cassandra: Use module based runner

... and

 - improve passthru.tests.
 - move version matrix into test module
 - remove rec keyword in package

roberth committed:
lib/modules: Add attrArgName

(cherry picked from commit d06f5fa5e4b579639531445daa68b4ef1ae0ad7a)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/specialargs-in-tests/19689/5

roberth committed:
nixos/testing/matrix: Add matrix choice to module args + format

roberth committed:
nixosTests.networking: Use module based runner

roberth committed:
nixos/testing: Remove redundant attributes from release.nix (hydra)

roberth committed:
nixos/testing/matrix: Add module indirection to avoid ambiguous `name` module arg

roberth committed:
Revert "lib/modules: Add attrArgName"

This reverts commit 1a896a3167037ddc0d0f7a73ee98dd898486b475.

roberth committed:
nixosTests.hostname: Fix warning

roberth committed:
lib/testing/matrix: matrix.*.value -> matrix.*.choice

roberth committed:
nixos/testing: Move entrypoint to nixos/lib + doc

roberth committed:
nixos/doc/writing-nixos-tests: Document how matrix options can be used

roberth committed:
doc/coding-conventions: Update Linking NixOS module tests to  package

ckiee commented on nixos/lib/testing/name.nix:
```diff
@@ -0,0 +1,7 @@
+{ lib, ... }:
+{
+  options.name = lib.mkOption {
+    description = "The name of the test.";
```
Was just scrolling through it to pass time and see the matrix comments, spotted this:
```suggestion
    description = "The name of the test";
```

ckiee replied:
Though then again, maybe nixpkgs has gotten big enough that the parts I touch have a slightly different style than the parts you touch. Who knows.

roberth replied:
If there's fewer than one sentence, it doesn't need a period. That makes sense, but the real problem is that I didn't write documentation for this option. The only info I added was "this option is part of the test somehow".

It's basically filler, like those cringy generated accessor docs you may find in some Java code.

roberth replied:
I mean I hope people didn't write that stuff :thinking:

roberth replied:
New doc:

> The name of the test.
>
> This is used in the derivation names of the [`driver`](#opt-driver) and [`test`](#opt-test) runner.

Seems like I've kept the period. Might make sense, depending on whether it's more of a sentence or more of a title. I think it's neither :shrug:.
I'll happily change it either way if it's the only thing this merge will depend on. I think we ought to fry bigger fish though.

roberth committed:
nixos/testing: Remove params

It is quite redundant.

roberth committed:
nixos/testing/matrix.nix: Only set a default for the module argument

roberth committed:
nixos/testing: Embrace callTest

My conception of its input was wrong. It is quite a useful construct,
even if its name is a bit weird.
The new exposeIntermediateAttrs function is the same pattern, but
applied to the intermediate nodes of the test tree, rather than the
leaves.

roberth committed:
nixos/tests: Add names

roberth commented:
All tests now pass, except for a handful of flaky ones: https://hydra.nixos.org/eval/1768916?compare=1768212&full=1
Note that some tests have been renamed. These can be found in the "Removed Jobs" and "New Jobs" tabs.
Some other tests haven't themselves been renamed, but are now instantiated correctly for both supported linux architectures and therefore have gained a suffix.

roberth committed:
nixos/doc/writing-nixos-test: Improve section about matrix module

roberth committed:
nixos/doc: Wire up the test options reference

roberth committed:
nixos/testing: Improve option docs

christian-burger commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
````diff
@@ -34,7 +37,64 @@ when switching between consoles, and so on. An interesting multi-node test is
+You can run a test with attribute name `mytest` in `nixos/tests/all-tests.nix` by invoking:
+
+```shell
+nix-build -A nixosTests.mytest
````
As a beginner, one sometimes misses the intricacies, that are very obvious to veteran maintainers. I might probably execute the given command to run tests at first in `nixpkgs/nixos/tests` (because I just edited `all-tests.nix` there) or `nixpkgs/nixos/` (because I added a test to NixOS), being befuddled by the error messages, before running it in `nixpkgs/` itself, where it actually works.

```suggestion
cd /my/git/clone/of/nixpkgs
nix-build -A nixosTests.mytest
```

roberth replied:
:+1: , thanks!

fricklerhandwerk commented on doc/contributing/coding-conventions.chapter.md:
````diff
@@ -652,7 +652,35 @@ stdenv.mkDerivation {
+```nix
+{ stdenv, lib, nixosTests }:
+
+# 1. finalAttrs.finalPackage is the correct package, even after `overrideAttrs`.
````
```suggestion
# 1. `finalAttrs.finalPackage` is the correct package, even after `overrideAttrs`.
```

fricklerhandwerk commented on doc/contributing/coding-conventions.chapter.md:
```diff
@@ -643,7 +643,7 @@ Here are examples of package tests:
 Like [package tests](#ssec-package-tests-writing) as shown above, [NixOS module tests](https://nixos.org/manual/nixos/stable/#sec-nixos-tests) can also be linked to a package, so that the tests can be easily run when changing the related package.
 
-For example, assuming we're packaging `nginx`, we can link its module test via `passthru.tests`:
+The basic way to set the `tests` attribute, if we're willing to make some assumptions, is the following. We use the `nginx` package as an example:
```
```suggestion
The basic way to set the `tests` attribute is the following.
We use the `nginx` package as an example:
```

Inserting that sentence makes the whole thing hard to parse. Also, you clarify the assumptions after the example, so it should be fine.

fricklerhandwerk commented on doc/contributing/coding-conventions.chapter.md:
````diff
@@ -652,7 +652,35 @@ stdenv.mkDerivation {
+```
+
+However, we're not guaranteed that `nixosTests.nginx` uses our package definition.
+Furthermore, some `nixosTests` have a test matrix that tests multiple versions.
+Let's look at `cassandra` for a more complete example:
````
```suggestion
Furthermore, some `nixosTests` have a test matrix that tests multiple versions.

Let's look at `cassandra` for a more complete example:
```

fricklerhandwerk commented on doc/stdenv/meta.chapter.md:
```diff
@@ -213,6 +213,8 @@ runCommand "my-package-test" {
 
 A timeout (in seconds) for building the derivation. If the derivation takes longer than this time to build, it can fail due to breaking the timeout. However, all computers do not have the same computing power, hence some builders may decide to apply a multiplicative factor to this value. When filling this value in, try to keep it approximately consistent with other values already present in `nixp ...
 
+This setting may be lost when the package is used as a dependency instead of being presented directly an evaluation process that handles the `meta.timeout` attribute. Meta attributes are not stored in the derivation.
```
```suggestion
`meta` attributes are not stored in the evaluated derivation.
Therefore, this setting may be lost when the package is used as a dependency.
To be effective, it must be presented directly to an evaluation process that handles the `meta.timeout` attribute.

```

fricklerhandwerk commented on nixos/doc/manual/development/running-nixos-tests-interactively.section.md:
```diff
@@ -33,3 +35,11 @@ $ ./result/bin/nixos-test-driver --keep-vm-state
+
+## Interactive-only test configuration {#sec-nixos-test-interactive-configuration}
+
+You can add configuration that is specific to the interactive test driver,
+by adding to the `interactive` option, which is copy of the regular test options
+namespace, which is used by the interactive test driver.
+This is useful for troubleshooting adaptations that you don't want to apply to
+regular test runs.
```
```suggestion
You can add configuration that is specific to the interactive test driver by adding to the `interactive` option.
`interactive` is a copy of the regular test options namespace, used by the interactive test driver.
This is useful for troubleshooting changes that you don't want to apply to regular test runs.
```

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
```diff
@@ -1,9 +1,9 @@
 # Writing Tests {#sec-writing-nixos-tests}
 
-A NixOS test is a Nix expression that has the following structure:
+A NixOS test is a Nix module that has the following structure:
```
```suggestion
A NixOS test is a NixOS module that has the following structure:
```

Oh boy, I just realized the NixOS manual _still_ does not have a proper definition of modules. The good one we have is [NixOS Wiki - Module](https://nixos.wiki/wiki/Module).

roberth replied:
I don't think we should call a module a NixOS module if it doesn't mix with actual NixOS modules. Calling it a "Nix module" (like I did) is also wrong. I'll use just "module".

fricklerhandwerk replied:
Good point.

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
````diff
@@ -21,10 +21,13 @@ import ./make-test-python.nix {
 ```
 
-The attribute `testScript` is a bit of Python code that executes the
+We refer to the whole test above as a test module, whereas the values
+in [`nodes.<name>`](#opt-nodes) are NixOS modules. (A NixOS configuration is a module.)
````
```suggestion
We refer to the whole test above as a test module, whereas the values
in [`nodes.<name>`](#opt-nodes) are NixOS modules themselves.
```

This should just link to a section that defines modules, once we have it. For now let's assume readers know what a NixOS module is, so we don't have to change this bit later.

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
```diff
@@ -21,10 +21,13 @@ import ./make-test-python.nix {
+We refer to the whole test above as a test module, whereas the values
+in [`nodes.<name>`](#opt-nodes) are NixOS modules. (A NixOS configuration is a module.)
+
+The option [`testScript`](#opt-testScript) is a bit of Python code that executes the
```
```suggestion
The option [`testScript`](#opt-testScript) is a piece of Python code that executes the
```

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
```diff
@@ -34,7 +37,65 @@ when switching between consoles, and so on. An interesting multi-node test is
-There are a few special NixOS configuration options for test VMs:
+## Calling a test {#sec-calling-nixos-tests}
+
+Tests are invoked a bit differently depending on whether the test lives in NixOS or in another project.
```
```suggestion
Tests are invoked differently depending on whether the test lives in NixOS or in another project.
```

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
```diff
@@ -34,7 +37,65 @@ when switching between consoles, and so on. An interesting multi-node test is
+
+### Testing within NixOS {#sec-call-nixos-test-in-nixos}
+
+Test modules can be instantiated into derivations in multiple ways.
```
```suggestion
Test modules can be added to derivations in multiple ways.
```

Not sure if "instantiate" is the right word at all.

fricklerhandwerk replied:
```suggestion
```

Actually this is stating the obvious, because the three variants follow immediately and there is nothing else in this section. We can remove it.

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
````diff
@@ -34,7 +37,65 @@ when switching between consoles, and so on. An interesting multi-node test is
+```
+
+Overrides can be added by defining an anonymous module in `all-tests.nix`.
+For the purpose of constructing a test matrix, [use the `matrix` options](#sec-nixos-test-matrix) instead.
````
```suggestion
For the purpose of constructing a test matrix, use the [`matrix` options](#sec-nixos-test-matrix) instead.
```

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
````diff
@@ -34,7 +37,65 @@ when switching between consoles, and so on. An interesting multi-node test is
+  hostname = runTest ./hostname.nix;
+```
+
+Overrides can be added by defining an anonymous module in `all-tests.nix`.
````
```suggestion
Overrides can be added by defining an anonymous module in `all-tests.nix` and passing it to `runTest`.
```

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
````diff
@@ -34,7 +37,65 @@ when switching between consoles, and so on. An interesting multi-node test is
+  hostname = runTest { imports = [ ./hostname.nix ]; defaults.networking.firewall.enable = false; };
+```
+
+You can run a test with attribute name `mytest` in `nixos/tests/all-tests.nix` by invoking:
+
+```shell
+cd /my/git/clone/of/nixpkgs
+nix-build -A nixosTests.mytest
````
It is a bit confusing that within the same example thread we use `hostname` and `mytest`.

```suggestion
You can run a test with attribute name `hostname` in `nixos/tests/all-tests.nix` by invoking:

```shell
cd /my/git/clone/of/nixpkgs
nix-build -A nixosTests.hostname
```

Alternatively use `mytest` consistently.

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
````diff
@@ -34,7 +37,65 @@ when switching between consoles, and so on. An interesting multi-node test is
+}
+```
+
+`runTest` returns a derivation that runs the test, or an attrset tree of such derivations.
````
```suggestion
`runTest` returns a derivation that runs the test, or an tree of attribute sets of such derivations.
```

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
````diff
@@ -34,7 +37,65 @@ when switching between consoles, and so on. An interesting multi-node test is
+For the purpose of constructing a test matrix, [use the `matrix` options](#sec-nixos-test-matrix) instead.
+
+```nix
+  hostname = runTest { imports = [ ./hostname.nix ]; defaults.networking.firewall.enable = false; };
````
For consistency with the other example, which is also multiline.

```suggestion
  hostname = runTest {
    imports = [ ./hostname.nix ];
    defaults.networking.firewall.enable = false;
  };
```

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
```diff
@@ -34,7 +37,65 @@ when switching between consoles, and so on. An interesting multi-node test is
+
+### Testing outside the NixOS project {#sec-call-nixos-test-outside-nixos}
+
+Outside the `nixpkgs` repository, you can instantiate the test by first acquiring the NixOS library,
```
```suggestion
Outside the `nixpkgs` repository, you can instantiate the test by first importing the NixOS library,
```

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
```diff
@@ -415,3 +475,120 @@ import ./make-test-python.nix
+`passthru.tests`. So instead, we recommend to use the `matrix` test options that
+take care of this for you.
+
+These options work by duplicating the entire test module configuration below
+the `matrix.<decision>.choice.<choice>.module` options.
```
How does order matter here?

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
````diff
@@ -415,3 +475,120 @@ import ./make-test-python.nix
+}
+```
+
+Instead of a single derivation, the test framework now returns:
````
```suggestion
Instead of a single derivation, the test framework transforms this to:
```

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
```diff
@@ -415,3 +475,120 @@ import ./make-test-python.nix
+
+It has picked up on the `matrix` and only returns the derivations constructed for each choice.
+
+The test related options such as `nodes` at the root of the test configuration still exist there, but are not demanded by `runTest`, so only their copies at each of the choices are evaluated. Hence, the "root" can see the consequences of the choices.
```
```suggestion
The test related options such as `nodes` at the root of the test configuration still exist there, but are not required by `runTest`, so only their copies at each of the choices are evaluated. Hence, the "root" can see the consequences of the choices.
```



> Hence, the "root" can see the consequences of the choices.


What does this mean? What does "see" mean? What are "consequences"?

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
````diff
@@ -415,3 +475,120 @@ import ./make-test-python.nix
+Some variations aren't simply parametric, but require specific configuration for some choices.
+These can be added in the `module` submodule:
+
+```nix
+{ lib, setup, ... }: {
+  matrix.backend.choice.smtp.module = { lib, setup, ... }: {
+    defaults.services.foo.backend = "smtp";
+    nodes.smtpserver = { … };
+    _module.args.setup = "";
+  };
+  matrix.backend.choice.rest.module = { lib, setup, ... }: {
+    _module.args.setup = ''
+      # extra setup for the rest backend
+    '';
+  };
+
+  nodes = …;
+  testScript = ''
+    ${setup}
+    # …
+  '';
+}
+```
````
This example requires more explanation to be useful. What is the role of `setup` here, and how is its value propagated through evaluation?

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
````diff
@@ -415,3 +475,120 @@ import ./make-test-python.nix
+}
+```
+
+And finally, some tests truly form a matrix, as they have two (or more) parameters, such as this example, producing the 6 test derivations that form the Cartesian product of `backend` and `testsuite`.
+
+Note that the modules inside `module` submodule can access their parent's module
+arguments through the lexical scope, where the matrix choice has not been decided
+yet. Any such parameters must be ignored or repeated inside the `module` submodule
+to get the final value.
+
+```nix
+{ backend, testsuite, ... }: {
+  matrix.backend.choice.smtp.module = { … };
+  matrix.backend.choice.rest.module = { … };
+
+  matrix.testsuite.choice = {
+    login-and-basics.module = {
+      testScript = …;
+    };
+    transactions.module = {
+      testScript = …;
+    };
+    advanced-use-cases.module = {
+      testScript = …;
+    };
+  };
+
+  name = "foo-${backend}-${testsuite}";
+
+  nodes.foo = …;
+}
+```
````
```suggestion
And finally, some tests truly form a matrix, as they have two (or more) parameters
The following example produces the 6 test derivations that form the Cartesian product of `backend` and `testsuite`.

Note that the modules inside `module` submodule can access their parent's module
arguments through the lexical scope, where the matrix choice has not been decided
yet. Any such parameters must be ignored or repeated inside the `module` submodule
to get the final value.

```nix
{ backend, testsuite, ... }: {
  matrix.backend.choice.smtp.module = { … };
  matrix.backend.choice.rest.module = { … };

  matrix.testsuite.choice = {
    login-and-basics.module = {
      testScript = …;
    };
    transactions.module = {
      testScript = …;
    };
    advanced-use-cases.module = {
      testScript = …;
    };
  };

  name = "foo-${backend}-${testsuite}";

  nodes.foo = …;
}
```
```

> Note that the modules inside `module` submodule can access their parent's module
arguments through the lexical scope, where the matrix choice has not been decided
yet. Any such parameters must be ignored or repeated inside the `module` submodule
to get the final value.

I really don't understand what this means. How exactly can the `module` submodules access their parent's argument? How would one get the final value? If this is important, please expand this note into a full example.

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
````diff
@@ -415,3 +475,120 @@ import ./make-test-python.nix
+}
+```
+
+And finally, some tests truly form a matrix, as they have two (or more) parameters, such as this example, producing the 6 test derivations that form the Cartesian product of `backend` and `testsuite`.
````
```suggestion
And finally, some tests truly form a matrix, as they have two (or more) parameters
The following example produces the 6 test derivations that form the Cartesian product of `backend` and `testsuite`.
```

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
```diff
@@ -415,3 +475,120 @@ import ./make-test-python.nix
+
+And finally, some tests truly form a matrix, as they have two (or more) parameters, such as this example, producing the 6 test derivations that form the Cartesian product of `backend` and `testsuite`.
+
+Note that the modules inside `module` submodule can access their parent's module
+arguments through the lexical scope, where the matrix choice has not been decided
+yet. Any such parameters must be ignored or repeated inside the `module` submodule
+to get the final value.
```
I really don't understand what this means. How exactly can the `module` submodules access their parent's argument? How would one get the final value? If this is important, please expand this note into a full example.

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
```diff
@@ -415,3 +475,120 @@ import ./make-test-python.nix
+  backend-rest = «derivation /nix/store/…-foo-rest.drv»;
+  backend-smtp = «derivation /nix/store/…-foo-smtp.drv»;
+
+  # make nix-build traverse this attrset
```
```suggestion
  # make `nix-build` traverse this attrset
```

fricklerhandwerk commented on nixos/lib/testing/matrix.nix:
```diff
@@ -0,0 +1,248 @@
+        description = ''
+          A attribute set where each attribute name corresponds to a choice; a single outcome of the decision.
+
+          When you define `matrix.foo.choice.bar`, the (new) module argument `foo` will be set to `bar` in a reinstantiation of the test configuration that produces the final test.
```
I don't unterstand this. How does `foo` become a (new - what does it mean to be new here?) argument? Can you show an example?

fricklerhandwerk commented on nixos/lib/testing/run.nix:
```diff
@@ -0,0 +1,57 @@
+      type = types.lazyAttrsOf types.raw;
+      description = mdDoc ''
+        Attributes to add to the returned derivations,
+        but are not necessarily part of the build.
```
```suggestion
        which are not necessarily part of the build.
```

fricklerhandwerk commented on doc/contributing/coding-conventions.chapter.md:
```diff
@@ -643,7 +643,7 @@ Here are examples of package tests:
 
```
This chapter should be in `nixos/doc`, as it is entirely NixOS-specific.

roberth replied:
I would agree that this aspect is at the interface between NixOS and Nixpkgs, but the _use case_ is entirely Nixpkgs-specific.
Package tests are a Nixpkgs concept. The implementation of such tests is quite often done by referring to NixOS tests, but that does not make them NixOS-specific. The focus is 1. the basics of linking those tests up, 2. dealing with version matrixes, which would otherwise make NixOS tests unsuitable.
I'm not entirely opposed to moving this, but it seems arbitrary. Did I miss something?

fricklerhandwerk replied:
Yeah, you're right. What really confuses me though is the cyclic dependency of `nixpkgs` on NixOS code here, but that is an unrelated issue.

fricklerhandwerk commented:
This is a fantastic addition. Albeit the PR is so large that I cannot imagine anyone seriously reviewing it in full. I only worked through the documentation part and only skimmed the code.

roberth committed:
nixos/doc/writing-nixos-tests: Various improvements

Thanks to fricklerhandwerk for the many suggestions, most of which
I have fixupped into preceding commits.

roberth commented:
<details><summary>Diff of fixups, which are about to disappear into a rebase in an attempt to have somewhat of an acceptable commit history. Thanks @fricklerhandwerk! :heart: </summary>

```diff
diff --git a/doc/contributing/coding-conventions.chapter.md b/doc/contributing/coding-conventions.chapter.md
index cf16ba88273..1e9ae075702 100644
--- a/doc/contributing/coding-conventions.chapter.md
+++ b/doc/contributing/coding-conventions.chapter.md
@@ -643,7 +643,9 @@ Here are examples of package tests:
 
 Like [package tests](#ssec-package-tests-writing) as shown above, [NixOS module tests](https://nixos.org/manual/nixos/stable/#sec-nixos-tests) can also be linked to a package, so that the tests can be easily run when changing the related package.
 
-The basic way to set the `tests` attribute, if we're willing to make some assumptions, is the following. We use the `nginx` package as an example:
+The basic way to set the `tests` attribute is the following.
+
+We use the `nginx` package as an example:
 
 ```nix
 { stdenv, lib, nixosTests }:
@@ -663,12 +665,13 @@ stdenv.mkDerivation {
 
 However, we're not guaranteed that `nixosTests.nginx` uses our package definition.
 Furthermore, some `nixosTests` have a test matrix that tests multiple versions.
+
 Let's look at `cassandra` for a more complete example:
 
 ```nix
 { stdenv, lib, nixosTests }:
 
-# 1. finalAttrs.finalPackage is the correct package, even after `overrideAttrs`.
+# 1. `finalAttrs.finalPackage` is the correct package, even after `overrideAttrs`.
 stdenv.mkDerivation (finalAttrs: {
   ...
 
diff --git a/doc/stdenv/meta.chapter.md b/doc/stdenv/meta.chapter.md
index ead9f27bb83..d3cfaba78af 100644
--- a/doc/stdenv/meta.chapter.md
+++ b/doc/stdenv/meta.chapter.md
@@ -213,7 +213,9 @@ runCommand "my-package-test" {
 
 A timeout (in seconds) for building the derivation. If the derivation takes longer than this time to build, it can fail due to breaking the timeout. However, all computers do not have the same computing power, hence some builders may decide to apply a multiplicative factor to this value. When filling this value in, try to keep it approximately consistent with other values already present in `nixpkgs`.
 
-This setting may be lost when the package is used as a dependency instead of being presented directly an evaluation process that handles the `meta.timeout` attribute. Meta attributes are not stored in the derivation.
+`meta` attributes are not stored in the instantiated derivation.
+Therefore, this setting may be lost when the package is used as a dependency.
+To be effective, it must be presented directly to an evaluation process that handles the `meta.timeout` attribute.
 
 ### `hydraPlatforms` {#var-meta-hydraPlatforms}
 
diff --git a/nixos/doc/manual/development/running-nixos-tests-interactively.section.md b/nixos/doc/manual/development/running-nixos-tests-interactively.section.md
index df3eb1d466e..6170057442d 100644
--- a/nixos/doc/manual/development/running-nixos-tests-interactively.section.md
+++ b/nixos/doc/manual/development/running-nixos-tests-interactively.section.md
@@ -38,8 +38,6 @@ directory.
 
 ## Interactive-only test configuration {#sec-nixos-test-interactive-configuration}
 
-You can add configuration that is specific to the interactive test driver,
-by adding to the `interactive` option, which is copy of the regular test options
-namespace, which is used by the interactive test driver.
-This is useful for troubleshooting adaptations that you don't want to apply to
-regular test runs.
+You can add configuration that is specific to the interactive test driver, by adding to the `interactive` option.
+`interactive` is a copy of the regular test options namespace, and is used by the interactive test driver.
+It can be helpful for troubleshooting changes that you don't want to apply to regular test runs.
diff --git a/nixos/doc/manual/development/writing-nixos-tests.section.md b/nixos/doc/manual/development/writing-nixos-tests.section.md
index e7406173ba5..94568bf3a12 100644
--- a/nixos/doc/manual/development/writing-nixos-tests.section.md
+++ b/nixos/doc/manual/development/writing-nixos-tests.section.md
@@ -1,6 +1,6 @@
 # Writing Tests {#sec-writing-nixos-tests}
 
-A NixOS test is a Nix module that has the following structure:
+A NixOS test is a module that has the following structure:
 
 ```nix
 {
@@ -22,9 +22,9 @@ A NixOS test is a Nix module that has the following structure:
 ```
 
 We refer to the whole test above as a test module, whereas the values
-in [`nodes.<name>`](#opt-nodes) are NixOS modules. (A NixOS configuration is a module.)
+in [`nodes.<name>`](#opt-nodes) are NixOS modules themselves.
 
-The option [`testScript`](#opt-testScript) is a bit of Python code that executes the
+The option [`testScript`](#opt-testScript) is a piece of Python code that executes the
 test (described below). During the test, it will start one or more
 virtual machines, the configuration of which is described by
 the option [`nodes`](#opt-nodes).
@@ -39,35 +39,36 @@ It uses two client nodes to test correct locking across server crashes.
 
 ## Calling a test {#sec-calling-nixos-tests}
 
-Tests are invoked a bit differently depending on whether the test lives in NixOS or in another project.
+Tests are invoked differently depending on whether the test is part of NixOS or lives in a different project.
 
 ### Testing within NixOS {#sec-call-nixos-test-in-nixos}
 
-Test modules can be instantiated into derivations in multiple ways.
-
 Tests that are part of NixOS are added to [`nixos/tests/all-tests.nix`](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/all-tests.nix).
 
 ```nix
   hostname = runTest ./hostname.nix;
 ```
 
-Overrides can be added by defining an anonymous module in `all-tests.nix`.
-For the purpose of constructing a test matrix, [use the `matrix` options](#sec-nixos-test-matrix) instead.
+Overrides can be added by defining an anonymous module in `all-tests.nix` and passing it to `runTest`.
+For the purpose of constructing a test matrix, use the [`matrix` options](#sec-nixos-test-matrix) instead.
 
 ```nix
-  hostname = runTest { imports = [ ./hostname.nix ]; defaults.networking.firewall.enable = false; };
+  hostname = runTest {
+    imports = [ ./hostname.nix ];
+    defaults.networking.firewall.enable = false;
+  };
 ```
 
-You can run a test with attribute name `mytest` in `nixos/tests/all-tests.nix` by invoking:
+You can run a test with attribute name `hostname` in `nixos/tests/all-tests.nix` by invoking:
 
 ```shell
 cd /my/git/clone/of/nixpkgs
-nix-build -A nixosTests.mytest
+nix-build -A nixosTests.hostname
 ```
 
 ### Testing outside the NixOS project {#sec-call-nixos-test-outside-nixos}
 
-Outside the `nixpkgs` repository, you can instantiate the test by first acquiring the NixOS library,
+Outside the `nixpkgs` repository, you can instantiate the test by first importing the NixOS library,
 
 ```nix
 # regular nix
@@ -91,7 +92,7 @@ nixos-lib.runTest {
 }
 ```
 
-`runTest` returns a derivation that runs the test, or an attrset tree of such derivations.
+`runTest` returns a derivation that runs the test, or a tree of attribute sets containing such derivations.
 
 ## Configuring the nodes {#sec-nixos-test-nodes}
 
@@ -493,7 +494,8 @@ the `matrix.<decision>.choice.<choice>.module` options.
 Laziness makes this pattern efficient.
 
 The test framework ([`runTest`](#sec-calling-nixos-tests)) takes care of
-traversing the choices.
+traversing the choices one by one, constructing a tree of attribute sets where
+each edge represents a choice.
 
 A parameterized test may look as follows:
 
@@ -516,14 +518,14 @@ A parameterized test may look as follows:
 }
 ```
 
-Instead of a single derivation, the test framework now returns:
+`runTest` transforms this to:
 
 ```nix
 {
   backend-rest = «derivation /nix/store/…-foo-rest.drv»;
   backend-smtp = «derivation /nix/store/…-foo-smtp.drv»;
 
-  # make nix-build traverse this attrset
+  # make `nix-build` traverse this attrset to evaluate nested tests
   recurseForDerivations = true;
   # …
 }
@@ -531,21 +533,27 @@ Instead of a single derivation, the test framework now returns:
 
 It has picked up on the `matrix` and only returns the derivations constructed for each choice.
 
-The test related options such as `nodes` at the root of the test configuration still exist there, but are not demanded by `runTest`, so only their copies at each of the choices are evaluated. Hence, the "root" can see the consequences of the choices.
+`runTest` works by invoking the module system to lazily produce all option values, and then extracts the relevant values from it.
+This means that some options exist, but are never evaluated. For example, when `matrix` isn't empty, `nodes.<name>` is ignored in favor of `matrix.<decision>.choice.<choice>.module.nodes.<name>`.
+
+Nonetheless, the definitions that would contribute to `nodes.<name>` are useful, as they contribute to `matrix.<decision>.choice.<choice>.module.nodes.<name>`. The `module` option always duplicates and extends the root of the test configuration.
 
 Some variations aren't simply parametric, but require specific configuration for some choices.
-These can be added in the `module` submodule:
+These can be added in the `module` submodule.
+
+The following example shows how tests in a test matrix can be quite different, while sharing common configuration.
 
 ```nix
 { lib, setup, ... }: {
-  matrix.backend.choice.smtp.module = { lib, setup, ... }: {
+  matrix.backend.choice.smtp.module = { lib, ... }: {
     defaults.services.foo.backend = "smtp";
     nodes.smtpserver = { … };
     _module.args.setup = "";
   };
-  matrix.backend.choice.rest.module = { lib, setup, ... }: {
+  matrix.backend.choice.rest.module = { lib, ... }: {
     _module.args.setup = ''
       # extra setup for the rest backend
+      # …
     '';
   };
 
@@ -557,12 +565,34 @@ These can be added in the `module` submodule:
 }
 ```
 
-And finally, some tests truly form a matrix, as they have two (or more) parameters, such as this example, producing the 6 test derivations that form the Cartesian product of `backend` and `testsuite`.
+The example shows the solutions to the following problems:
+ - The test cases require different configuration. For example, the `smtp` test sets the backend to `"smtp"` for all nodes, instead of relying on the default value.
+ - One test case requires an extra VM. `nodes.smtpserver` is only defined for the `smtp` test.
+ - One test case needs extra setup to be done in the `testScript`. It uses an ad hoc module argument called `setup` to achieve this.
+
+You may notice that the test appears to pull a `setup` value out of thin air, and indeed this would fail if `runTest` had accessed `testScript` directly instead of looking inside the `matrix.<...>.module` options. Thanks to laziness, this is not a problem.
+
+You may also notice that there was no use of the `backend` module argument anymore, as all decisions' effects were defined in the `matrix.<...>.module` options. You could re-add the `backend` to the parameter list to combine both styles.
+
+If you have many test cases, but few non-empty `setup` values, you could define a default value for the module argument instead:
+
+```nix
+{ lib, setup, ... }: {
+  _module.args.setup = lib.mkDefault "";
+  matrix.backend.choice.smtp.module = { lib, ... }: {
+    defaults.services.foo.backend = "smtp";
+    nodes.smtpserver = { … };
+  };
+  # …
+}
+```
+
+If you are writing a reusable module, you may declare a regular option instead, using `lib.mkOption`. <!-- TODO refer to a "reusable modules" section in module system docs (to be written) -->
 
-Note that the modules inside `module` submodule can access their parent's module
-arguments through the lexical scope, where the matrix choice has not been decided
-yet. Any such parameters must be ignored or repeated inside the `module` submodule
-to get the final value.
+### Multiple decisions {#sec-nixos-test-matrix-multiple-decisions}
+
+And finally, some tests truly form a matrix, as they have two (or more) parameters.
+The following example produces the 6 test derivations that form the Cartesian product of `backend` and `testsuite`.
 
 ```nix
 { backend, testsuite, ... }: {
@@ -587,6 +617,40 @@ to get the final value.
 }
 ```
 
+### Ordering and caveats {#sec-nixos-test-matrix-ordering}
+
+It is usually sufficient to think only about the final test configuration where all decisions have been made, but it is possible to access non-final test configurations, on purpose or by accident.
+
+This possibility is useful when the availability of some choices depends on previous choices, to construct something akin to a decision tree rather than a matrix. See [`matrix.<name>.enable`](#opt-matrix._name_.enable).
+
+However, it can also be unexpected, when you use a "free variable" in `matrix.<...>.module`; a variable from the lexical scope, rather module arguments declared by the expression following `module =`. For example:
+
+```nix
+{ y, ... }: {
+  matrix.x.choice.foo.module =
+    # the following expression has `y` as a free variable:
+    {
+      # this `y` does not come from the final configuration, but rather from
+      # an incomplete configuration where `x` has not been decided yet.
+      bar = y;
+    };
+  matrix.y = …;
+}
+```
+
+The dependency on `y` can be fixed by declaring it as a module argument, so that it comes from the final test configuration.
+
+```nix
+{ ... }: {
+  matrix.x.choice.foo.module = { y, ... }: {
+    bar = y;
+  };
+  matrix.y = …;
+}
+```
+
+You could also fix it by [ordering](See [`matrix.<decision>.after`](#opt-matrix._name_.after).) `x` to be decided after `y`. This is only preferable for dependencies that aren't in `matrix.<...>.module`.
+
 ## Test Options Reference {#sec-test-options-reference}
 
 The following options can be used when writing tests.
diff --git a/nixos/lib/testing/matrix.nix b/nixos/lib/testing/matrix.nix
index 97a2eb4ed76..fe1145a901a 100644
--- a/nixos/lib/testing/matrix.nix
+++ b/nixos/lib/testing/matrix.nix
@@ -58,7 +58,9 @@ let
         description = ''
           A attribute set where each attribute name corresponds to a choice; a single outcome of the decision.
 
-          When you define `matrix.foo.choice.bar`, the (new) module argument `foo` will be set to `bar` in a reinstantiation of the test configuration that produces the final test.
+          When you define `matrix.foo.choice.bar`, it creates a copy of the test configuration inside `matrix.foo.choice.bar.module`, where the module argument `foo` will be set to `bar`.
+
+          `runTest` will traverse all matrix choices, so that in all the final test configurations it returns, a choice for `foo` has been made.
         '';
       };
     };
diff --git a/nixos/lib/testing/run.nix b/nixos/lib/testing/run.nix
index d28eba425b3..0cd07d8afd2 100644
--- a/nixos/lib/testing/run.nix
+++ b/nixos/lib/testing/run.nix
@@ -8,7 +8,7 @@ in
       type = types.lazyAttrsOf types.raw;
       description = mdDoc ''
         Attributes to add to the returned derivations,
-        but are not necessarily part of the build.
+        which are not necessarily part of the build.
 
         This is a bit like doing `drv // { myAttr = true; }` (which would be lost by `overrideAttrs`).
         It does not change the actual derivation, but adds the attribute nonetheless, so that
         
```

</details>

roberth committed:
nixos/testing/matrix: Remove movie references

Except for one easter egg decorating an error message that should
be easy to resolve.

roberth commented:
> This is a fantastic addition.

:pray: 

> Albeit the PR is so large that I cannot imagine anyone seriously reviewing it in full.

It is a lot of stuff, but what may save it is:
 - I've taken great care not to change the logic
 - this has run on hydra to make sure we don't have regressions (successfully, and will do again before merge or after touching the logic again)
 - I prioritize aftercare

> to have a guide how to roughly perform evaluation in your head, because all of the examples read like magic. Right now I would have to dive through code to understand what they actually do and what they evaluate to. I would not be confident at all to start writing my own tests. Maybe a brief documentation of the test framework architecture would be in order, as an introductory subsection.

Except for summarizing the evaluation, I think I might have already improved the situation a bit by touching on what gets evaluated in the matrix option descriptions.
I could still add an explanation of how it evaluates, but that would be a bit too detailed to serve as an introduction I think.

> Ideally you would sit down with someone and watch them using the documentation to write a test on their own, it will reveal the most pressing problems with the explanations.

That would be ideal, but I don't have a lot of budget for this kind of thing yet. Will focus on this for the features (matrix options), but not do the whole test docs.

Fwiw, delaying the code change also has a human cost.

fricklerhandwerk commented on nixos/lib/testing/matrix.nix:
```diff
@@ -0,0 +1,255 @@
+        This option constructs a hierarchy of options, where each decision to be made is represented by `<...>.matrix.<name>`.
+        These options themselves contain `choice.<name>`, which represent the possible values for each decision.
+
+        The `module` submodule is special, because it duplicates and extends the entire test configuration the existed before making the decision.
```
```suggestion
        The `module` submodule is special, because it duplicates and extends the entire test configuration that existed before making the decision.
```

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
```diff
@@ -417,3 +478,183 @@ import ./make-test-python.nix
+
+It has picked up on the `matrix` and only returns the derivations constructed for each choice.
+
+`runTest` works by invoking the module system to lazily produce all option values, and then extracts the relevant values from it.
```
```suggestion
`runTest` works by invoking the module system to lazily produce all option values.
```

I think the next sentence says the same thing that I removed.

roberth replied:
I want to stress here that some option values go unused because of laziness and how the matrix module works.
"for each test configuration" seems implied and does not describe how the code works, even if it's somewhat true at a high level.

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
```diff
@@ -417,3 +478,183 @@ import ./make-test-python.nix
+It has picked up on the `matrix` and only returns the derivations constructed for each choice.
+
+`runTest` works by invoking the module system to lazily produce all option values, and then extracts the relevant values from it.
+This means that some options exist, but are never evaluated. For example, when `matrix` isn't empty, `nodes.<name>` is ignored in favor of `matrix.<decision>.choice.<choice>.module.nodes.<name>`.
```
```suggestion
This means that some options exist, but are never evaluated. For example, when `matrix` is set, `nodes.<name>` is ignored in favor of `matrix.<decision>.choice.<choice>.module.nodes.<name>`.
```

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
```diff
@@ -417,3 +478,183 @@ import ./make-test-python.nix
+`runTest` works by invoking the module system to lazily produce all option values, and then extracts the relevant values from it.
+This means that some options exist, but are never evaluated. For example, when `matrix` isn't empty, `nodes.<name>` is ignored in favor of `matrix.<decision>.choice.<choice>.module.nodes.<name>`.
+
+Nonetheless, the definitions that would contribute to `nodes.<name>` are useful, as they contribute to `matrix.<decision>.choice.<choice>.module.nodes.<name>`. The `module` option always duplicates and extends the root of the test configuration.
```
```suggestion
Nonetheless, the definitions in `nodes.<name>` are useful, as they contribute to `matrix.<decision>.choice.<choice>.module.nodes.<name>`.
The `module` option always duplicates and extends the root of the test configuration.
```

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
```diff
@@ -417,3 +478,183 @@ import ./make-test-python.nix
+ - One test case requires an extra VM. `nodes.smtpserver` is only defined for the `smtp` test.
+ - One test case needs extra setup to be done in the `testScript`. It uses an ad hoc module argument called `setup` to achieve this.
+
+You may notice that the test appears to pull a `setup` value out of thin air, and indeed this would fail if `runTest` had accessed `testScript` directly instead of looking inside the `matrix.<...>.module` options. Thanks to laziness, this is not a problem.
```
This does not explain at all why it works.

```suggestion
You may notice that the test appears to pull a `setup` value out of thin air.
Indeed, this would fail if `runTest` had accessed `testScript` directly instead of looking inside the `matrix.<...>.module` options.
Instead, `runTest` calls the top-level test module declaration in each of the configurations it produces, passing the `setup` argument with the value set in `_module.args.setup`.
```

Is this correct?

roberth replied:
A correct understanding, but with some incorrect technicalities.
 - "declaration" should be "definition"
 - we're giving `runTest` a bit too much agency, but if we need to mention all indirections, we're not more helpful than the code itself. This applies also to my own writing as well.

I'm changing it to this

```suggestion
You may notice that the test appears to pull a `setup` value out of thin air.
Indeed this would fail if `runTest` had accessed `testScript` directly instead of looking inside the `matrix.<...>.module` options.
In those `matrix.<...>.module` contexts, the definition of `testScript` is valid.
Thanks to laziness, the invalid `testScript` is never accessed and never becomes a problem.
```

fricklerhandwerk commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
````diff
@@ -417,3 +478,183 @@ import ./make-test-python.nix
+}
+```
+
+It has picked up on the `matrix` and only returns the derivations constructed for each choice.
````
```suggestion
It has picked up on the `matrix` and only returns the derivations constructed for each choice.

The arguments to the test module, which is called by `runTest` to create each configuration in the matrix, are built up as follows:
`matrix.<decision>` corresponds to the argument name to the test module.
`choice.<choice>` corresponds to the named argument's value as a string.
```

Something along these lines would have made clear to me from the start how the attributes are related to the test module arguments.

roberth replied:
Changing to

```suggestion
It has picked up on the `matrix` and only returns the derivations constructed for each choice.

The arguments to the test module, which is called by `runTest` to create each configuration in the matrix, are built up as follows:
 - `<decision>` in `matrix.<decision>` corresponds to the argument name to the test module.
 - `<choice>` in `choice.<choice>` corresponds to the named argument's value as a string. Non-string values can be set with `_module.args` or other options inside [`choice.<choice>.module`](#opt-matrix._name_.choice._name_.module).

```

fricklerhandwerk reviewed (commented):
Yes, I surely do not want to stall this PR. Yet let me throw in the consideration that having sub-par documentation will perpetuate the situation we currently have with people being confused to the point of not being able to use features like these. The recent changes appear to me to have the same problem as @Ericson2314's cross compilation notes: they are deeply cursed by intimate knowledge of the implementation. I tried to address this with suggestions.

I won't oppose merging, but would like to note that I fear it may be more work for someone else, not involved here, to sit down later and figure out how to simplify this than discussing it while it's still warm.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2022-08-01-summer-of-nix-documentation-q-a/20695/1

fricklerhandwerk reviewed (approved):
LGTM in terms of docs.

roberth commented on nixos/lib/testing/matrix.nix:
```diff
@@ -0,0 +1,255 @@
+  };
+
+  choiceModule = decisionName: extendModules: { name, ... }: {
+    options = {
```
Maybe

```suggestion
    options = {
      enable = ...
```

An easier way to filter out nonsensical/unsupported combinations.

roberth replied:
Done.

roberth committed:
nixos/doc/writing-nixos-tests: Using the right scope

roberth committed:
nixos/testing: Rename local variable

roberth committed:
nixos/testing: Add matrix.<name>.choice.<name>.enable

fricklerhandwerk commented on nixos/doc/manual/development/running-nixos-tests-interactively.section.md:
```diff
@@ -38,6 +38,12 @@ directory.
-You can add configuration that is specific to the interactive test driver, by adding to the `interactive` option.
-`interactive` is a copy of the regular test options namespace, and is used by the interactive test driver.
-It can be helpful for troubleshooting changes that you don't want to apply to regular test runs.
+The `.driverInteractive` attribute combines the regular test configuration with
+definitions from the [`interactive` submodule](#opt-interactive). This gives you
+a more usable, graphical, but slightly different configuration.
+
+To run only the regular configuration, build the `<test>.driver` attribute
+instead, and call it with the flag `result/bin/nixos-test-driver --interactive`.
+
+You can add your own interactive-only test configuration by adding extra
+configuration to the [`interactive` submodule](#opt-interactive).
```
Not sure if I read the code correctly, but `driverInteractive` seems to be only used internally. If that's true it should not even be mentioned here.

```suggestion
You can add your own interactive-only test configuration by adding extra
configuration to the [`interactive` submodule](#opt-interactive).

It will be combined with the regular test configuration, and provides a more convenient, graphical test environment.

To run interactvely the regular configuration, build the `<test>.driver` attribute
instead, and call it with the `interactive` flag. For example:

    result/bin/nixos-test-driver --interactive


```

roberth replied:
`.driverInteractive` is specifically for troubleshooting the user's tests, so it is a public interface.

So there's three ways to run a test
 - regular configuration, non-interactive
 - regular configuration, interactive
 - interactive configuration, interactive

I considered adding this table, but I find the cell contents too vague.

|                                           | regular mode              | `--interactive` mode                                      |
|-------------------------------------------|---------------------------|-----------------------------------------------------------|
| regular configuration                     | `nix-build --no-out-link` | `nix-build -A nixosTests.<name>.driver` + `--interactive` |
| graphical and `interactive` configuration | not useful                | `nix-build -A nixosTests.<name>.driverInteractive`        |

The underlying problem is that we have two uses of "interactive" that aren't quite the same.
 - interactive configuration: a graphical qemu among other things
 - running the driver interactively: with `--interactive`, which is also implied by `.driverInteractive`.

I'd like for this to be improved, but this is all pre-existing and I'll consider it out of scope for this PR.

fricklerhandwerk replied:
> So there's three ways to run a test

This was clear to me from the description. What's not at all clear how `driverInteractive` plays into this and how one would, for example as you say, troubleshoot setups.

roberth replied:
You'd follow the instructions and then you get a python repl where you can do anything the `testScript` can do, as described in `#sec-running-nixos-tests-interactively`.

> how driverInteractive plays into this

It seems that this paragraph answers your question:

> The `.driverInteractive` attribute combines the regular test configuration with
> definitions from the [`interactive` submodule](#opt-interactive). This gives you
> a more usable, graphical, but slightly different configuration.

Could you perhaps refine your question?

fricklerhandwerk commented on nixos/lib/testing/interactive.nix:
```diff
@@ -6,11 +6,31 @@ in
     interactive = mkOption {
       description = mdDoc ''
-        Tests [can be run interactively](#sec-running-nixos-tests-interactively).
+        Tests [can be run interactively](#sec-running-nixos-tests-interactively)
+        using the program in the test derivation's `.driverInteractive` attribute.
```
As above, this seems to unnecessarily expose internals.

```suggestion
        Tests [can be run interactively](#sec-running-nixos-tests-interactively).
```

roberth replied:
`.driverInteractive` is crucial for using the `interactive` submodule, so it is worth repeating here.

`sec-running-nixos-tests-interactively` documents the user interface. `.driverInteractive` isn't just part of the UI but part of `interactive`'s _relation to_ the UI, so it should be mentioned here alongside the link.

fricklerhandwerk replied:
Ah now I found it in the manual section that was not part of the diff. Sorry for the noise.

tfc commented on nixos/lib/testing/driver.nix:
```diff
@@ -0,0 +1,188 @@
+      description = mdDoc ''
+        Extra arguments to pass to the test driver.
+
+        The become part of [{option}`driver`](#opt-driver) via `wrapProgram`.
```
```suggestion
        They become part of [{option}`driver`](#opt-driver) via `wrapProgram`.
```

roberth committed:
nixos/testing: Add matrix.<decision>.choice.<choice>.value alias

roberth committed:
nixos/testing: Add matrix.<decision>.isBool convenience

roberth committed:
nixos/testing: Rename <choice>.module -> <choice>.extraConfig

roberth commented:
Regarding the build status: I've fixed the manual, but github actions is checking out the wrong commit.
Next step is finalizing review and testing, so I'll ignore it for now.

olafklingt commented:
Would it make sense to separate reference documentation from guides?
Reference is for example https://nixos.org/manual/nixos/stable/index.html#ssec-machine-objects
But also configuration of VMs and the test options section.

fricklerhandwerk commented:
> Would it make sense to separate reference documentation from guides?

@olafklingt yes, and we should still merge this as is. We can always move around contents later. PRs should be as small as possible to get any changes in at all, otherwise it's too much of a burden for both authors and reviewers.

olafklingt commented on nixos/doc/manual/development/writing-nixos-tests.section.md:
```diff
@@ -21,10 +21,13 @@ import ./make-test-python.nix {
 
-The attribute `testScript` is a bit of Python code that executes the
+We refer to the whole test above as a test module, whereas the values
+in [`nodes.<name>`](#opt-nodes) are NixOS modules themselves.
```
I think it makes more sense to not use the term "NixOS module" but "NixOS configuration" because NixOS configurations (as they are first seen by users in the `configuration.nix` are a more specific type of function than modules, they just make use of the module convention.

roberth commented:
I've separated out the modularization bits into https://github.com/NixOS/nixpkgs/pull/191540 which I've tested successfully on Hydra and is ready for review.

Atemu commented:
How do we want to continue on this? We must not let this bit rot. It's a much needed improvement.

The release and therefore feature freezes are coming up. I don't think it'd be good to merge such a change this late in the cycle.

Assuming no major faults are found, do we want to merge this right after ZHF ends/the release?

roberth commented:
I don't think I'll have enough time to do it before 22.11, but I don't want to let it rot either. Aspirationally, I want to finish this around December or January.

Atemu commented:
Hi, what's the status on this?

roberth committed:
nixos/lib/testing: Pull callTest in

It makes more sense to let the modules guide the invocation of this
function rather than to produce a data structure and then consume it
immmediately.

Maybe the same should be done for exposeIntermediateAttrs. It'd turn
the two functions into a catamorphism that's integrated into the
modules.

roberth commented:
@Atemu just rebased. Annoying but wasn't all that bad. Personally I can't spend a lot of time on this, but it seems that the PR is in a decent state now.

TODO: hacking
 - [ ] maybe refactor `exposeIntermediateAttrs` as mentioned in last commit: (nixos/lib/testing: Pull callTest in)
 - [ ] `extraConfig` -> `test` <= just use the term from the domain. It's a test configuration module, so call it `test`.

TODO: review
 - [ ] thorough review of the correctness of the rebase, especially check for subtle changes in the tests
 - [ ] final review of docs

 > Maybe the same should be done for exposeIntermediateAttrs. It'd turn
> the two functions into a catamorphism that's integrated into the
> modules.

TODO: final double-check; manual testing
- [ ] `nixosTests`
- [ ] `nixosTest`
- [ ] `nixos.lib.*`
- [ ] `testing-python.nix`/`makeTest`
- [ ] `nixos/release-combined.nix`/`nixos.tests`

Atemu commented:
I think we should do our best to get an MVP of this PR without major regressions in first. Any further niceties and minor refactorings can come later. It's important further changes to tests can make use of this PR's core changes ASAP so that we don't diverge endlessly.   

Would a hydra jobset to verify help?

roberth commented:
Just need to do a bit of due diligence, but yeah, no need to drag this out.

> Would a hydra jobset to verify help?

There's a jobset any nixpkgs maintainer can reuse: https://hydra.nixos.org/jobset/nixos/python-test-refactoring. I'll document how to use it effectively.

roberth commented:
> Would a hydra jobset to verify help?

Documenting the testing process here https://github.com/NixOS/nixpkgs/pull/216660

roberth commented:
I have to be honest: I don't think I'll get around to completing this. It'd be great if someone could take over.

Also: don't use the existence of this as an excuse not to write parameterized nixos tests. The true cost of writing a test is in finding the requirements and doing implementation work. Changing an already parameterized test to use a framework like this is tiny.

roberth commented:
This needs modularization.

In the current code, it adds a hook to override the return value. This isn't terrible, but obscures the fact that conceptually the test matrix comes *before* the test, even if both things can be specified in the same module. It also makes it impossible to use the test matrix logic in a context other than the NixOS test framework, or in a combined way.

TODO:
 - Simplify
   - [ ] One option / attr to retrieve `release.nix` attributes (Hydra)
   - [ ] One option / attr to retrieve `nix-build` attributes (also OfBorg and `nix build` with `legacyPackages`)
   - [ ] One option / attr to retrieve flattened attributes (e.g. flake `checks`)
 - Extraction
   - [ ] Factor out the matrix module. I think this is almost done thanks to it being a separate file
   - [ ] Document the matrix system
   - [ ] Test the matrix evaluation independent of the rest of the test framework
 - Integration
   - [ ] Use it with the test framework. `imports` the test framework into it?
   - [ ] Use it with simple build tests, such as the `etc` tests
   - [ ] Update all-tests.nix to use it
   - [ ] Apply it to the entire all-tests.nix? A test suite is just a test matrix where one of the axes is a set of test cases. (And the matrix (or tensor) does not have to be rectangular, boxy, or whatever math term we want to abuse for it. Not all coordinates have to exist. The shape can be irregular. Etc etc.)

AleXoundOS commented:
Is it true that without merging this PR currently
one cannot override arbitrary parts of the `nodes` attribute set
of a test from nixos/tests?

<details> 
  <summary>My attempts have failed. </summary>
   If I import `nixos/tests/some-test.nix`, the resulting set is not suitable for reuse in `runTest`.
NixOS tests import `make-test-python.nix` right on the first line, that doesn't allow to access the original `nodes` attribute set.
</details>
